src/Polly.Core/ResilienceProperties.cs

```diff
@@ -20,10 +20,18 @@ public sealed class ResilienceProperties
     /// <returns>True, if a property was retrieved.</returns>
     public bool TryGetValue<TValue>(ResiliencePropertyKey<TValue> key, [MaybeNullWhen(false)] out TValue value)
```

Comment:
This would be a breaking change though.

To me the main question is this -> is this operation legal in the context of ResilienceProperties?
```cs
props.Set(key, null);
```

If yes - then the annotation is incorrect.
Thought to me this feels like a request to remove an item from the collection...

Reply:
> Assuming a null value is permissible in the first place, doesn't that mean this annotation is wrong?

As I understand, the attribute we are currently using `[MaybeNullWhen(false)]` is the best option here, because it ensures better null-state static analysis. For example, currently the static analysis can correctly analyze the following scenarios:

- `TValue` is non-nullable reference type: `a` - is not null and we can declare it as `string`, while `b` is nullable and must be declared as `string?`:
![{B090FA96-4768-4CF4-A4C8-6DFBDCF35744}](https://github.com/user-attachments/assets/e6996626-4b71-482f-b427-f5f7e489cb25)
- `TValue` is nullable reference type: `a` and `b` nullable, hence must be declared as `string?`:
![{EEC850C2-E27D-4937-B3D6-2CF465A91FC5}](https://github.com/user-attachments/assets/a355ffdd-7faa-4542-ac09-08c7090e8309)

We can try and remove the attribute or replace it with `[MaybeNull]` attribute but that would give us worse static analysis results. For example, if use `[MaybeNull]` then the first sample mentioned above will produce the following result:
![{303B4BB0-14C8-4DE1-9D45-4471103881F4}](https://github.com/user-attachments/assets/93d52a3c-1694-4158-923e-7f6da9ebd15a)
This result is not correct because we declared `TValue` as non-nullable `string`, and hence `a` must have type `string`, but the analyzer requires it to be `string?`.

For example, `System.Collections.Generic.Dictionary` uses the same attribute as we do for `TryGetValue` method, see [here](https://github.com/dotnet/runtime/blob/b5833d2cd22f474485f74696609f6a6c3617cd15/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1441), although the values can be null.

To conclude, I think the attribute is correct, but in one particular case we have to use null-forgiving operator in the implementation to satisfy the analyzer.

Reply:
> To me the main question is this -> is this operation legal in the context of ResilienceProperties?
> ```csharp
> props.Set(key, null);
> ```

I don't know. There could be a value for users if they treat `null` as a special value they want to handle in a specific way, but I'm not aware of any real scenarios.

Reply:
@martintmk Could you help here, please? When you were adding `ResilienceProperties` was it assumed that the `props.Set(key, null)` operation would be legal?

Reply:
> Could you help here, please? When you were adding ResilienceProperties was it assumed that the operation props.Set(key, null) would be legal?

To me, yes, there are scenarios where users can explicitly set a null to a resilience property and treat it as valid value. I don't see why we should have this flexibility.

Reply:
Then I would keep `[MaybeNullWhen(false)]` attribute, since it ensures better null state static analysis.

@martincostello @RussKie What do you think?

Reply:
I'm happy provided that the compiler treats the value appropriately in the case where it returns `true` (i.e. you get warned the value might be `null`).

Reply:
Fine for me too, given it does not produce "noise warnings".

Reply:
If `null` is a valid value, then the contract has to be
```cs
bool TryGetValue<TValue>(ResiliencePropertyKey<TValue> key, [MaybeNullWhen(true)] out TValue value)
```

...and this is a behavioural change.

Reply:
> I'm happy provided that the compiler treats the value appropriately in the case where it returns `true` (i.e. you get warned the value might be `null`).

The warning will be produced if `TValue` is a nullable reference type, e.g. `string?`. In this case we expect values to be nullable, hence we want the warning to be generated.

In case when `TValue` is non-nullable, e.g. `string`, the warning will not be generated. And if a users decides to put null values into resilience properties (regardless the fact that `TValue` is non-nullable), the analyzer will not warn him/her about possible null values returned by `GetValue` and `TryGetValue` methods.

Reply:
@RussKie 
`[MaybeNullWhen(true)]` - if we use this then we'll instruct the analyzer that the value is always **not-null** when the method returns `false`, which is not correct.

![image](https://github.com/user-attachments/assets/dc7f4151-e80e-491d-839a-454165a23b94)

Reply:
Ah, yes. Then the attribute should be removed completely.

Reply:
Removing the attribute will force us to use `!` in two places:
![{FD895A68-AB4B-47FB-BEA5-CB53EBBFD65A}](https://github.com/user-attachments/assets/40217dcf-6663-420f-a12f-ff109bcf7c6a)

And will worse the results of the static analysis, e.g.:
![{CFE36ACD-A06B-4857-A25B-6CDB1B4CAD84}](https://github.com/user-attachments/assets/848a704c-b6a3-48fd-ada2-ef35fe0a61b4)
![image](https://github.com/user-attachments/assets/95ac0fce-d3f2-4594-886a-c6d9cbe7912f)

So, what we currently have is the best option available, and we have to use `!` in that one assignment to satisfy the analyzer.

Reply:
There are no changes to the signature of this method. I think the analyzer adds the value. I am fine merging this as it is. cc @martincostello ?

Reply:
I agree. The PR is good to be merged.

Reply:
```cs
bool TryGetValue<TValue>(ResiliencePropertyKey<TValue> key, [MaybeNullWhen(false)] out TValue value)
```

Let's step back and look at it from the consumer angle. `value` can be `null` irrespective of the return value. Which means the current annotation `[MaybeNullWhen(false)]` is incorrect, because it implies it's never `null` when the return value is `true`. This violates the contract and will, essentially, result in NRE in the consumer code (and can/will crash apps):
```cs
            else if (val is null)
            {
                value = default!;
```

Reply:
If valuetypes are expected, then there has to be two methods:
![image](https://github.com/user-attachments/assets/2c9ebe63-5e37-4275-bb1c-e2b74bc01d89)

If not, then one would suffice.

Reply:
> If valuetypes are expected, then there has to be two methods: 
> ...
> If not, then one would suffice.

We do want value-types. The sample you provided doesn't cover cases when `out T` is no-nullable reference-type and value-type. If you try to add support of those to your sample you'll end up using `!` to suppress analyzer's warning when you want to return `null`.

> Let's step back and look at it from the consumer angle. `value` can be `null` irrespective of the return value. Which means the current annotation `[MaybeNullWhen(false)]` is incorrect, because it implies it's never `null` when the return value is `true`. This violates the contract and will, essentially, result in NRE in the consumer code (and can/will crash apps):

[Documentation](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.maybenullwhenattribute?view=net-8.0) says `the parameter may be null even if the corresponding type disallows it`. It means that this annotation is useful when `T` is non-nullable, e.g. a customer users `string` as `T`, and in one particular case when `ReturnValue=false` a value of `T` maybe `null`. But when `T` is nullable, e.g. `string?`, the annotation doesn't bring any value, because `string?` is nullable hence its value can be `null` regardless of the `ReturnValue`.

Reply:
We discussed the issue with @RussKie and agreed that we are good with `[MaybeNullWhen(false)]` attribute. I'll add a comment explaining the need to use `!` operator.

Reply:
I'll aim to take a proper look at this today once that's pushed.

Reply:
Thank you @iliar-turdushev  for your patience and explaining the reasons.